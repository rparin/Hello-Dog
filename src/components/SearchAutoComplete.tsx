"use client";
import React, { useState } from "react";
import { ReactSearchAutocomplete } from "react-search-autocomplete";
import { dogItem, dogBreedList } from "@data/DogBreedList";
import { useRouter } from "next/navigation";

export default function SearchAutoComplete({
  className,
  placeholder,
  autofocus,
}: {
  className?: string;
  autofocus?: boolean;
  placeholder?: string;
}) {
  // Hold search values
  const [searchValue, setSearch] = useState<undefined | string>(undefined);
  const router = useRouter();
  const handleOnSelect = (item: dogItem) => {
    router.push(`/${item.name.replace(/ /g, "_")}`);
  };
  const handleOnSearch = (value: string, results: dogItem[]) => {
    if (value.replace(/ /g, "") != "" && results.length > 0) {
      setSearch(results[0].name);
    }
  };

  const handleOnSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (searchValue) {
      router.push(`/${searchValue.replace(/ /g, "_")}`);
    }
  };

  return (
    <form onSubmit={handleOnSubmit} className={className}>
      <ReactSearchAutocomplete
        className="w-full"
        items={dogBreedList}
        onSelect={handleOnSelect}
        onSearch={handleOnSearch}
        // formatResult={(result: dogItem) => {
        //   return <>{result.name.replace(/_/g, "-")}</>;
        // }}
        styling={{
          border: "2.5px solid #FFD29D",
          lineColor: "rgba(0,0,0,.35)",
          boxShadow: "rgba(32, 33, 36, 0.28) 0px 1px 12px 0px",
          fontFamily: "Belanosima, Franklin Gothic Medium, Arial Black",
        }}
        placeholder={placeholder}
        showIcon={false}
        maxResults={5}
        autoFocus={autofocus}
      />
    </form>
  );
}
