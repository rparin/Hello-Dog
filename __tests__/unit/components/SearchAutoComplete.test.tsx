import "@testing-library/jest-dom";
import { act, fireEvent, render, screen } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import SearchAutoComplete from "@/components/SearchAutoComplete";

describe("testing SearchAutoComplete Accessibility", () => {
  expect.extend(toHaveNoViolations);

  test("it should have no accessibility violations", async () => {
    const { container } = render(<SearchAutoComplete />);
    const results = await act(async () => axe(container));
    expect(results).toHaveNoViolations();
  });
});

test("it renders component unchanged", () => {
  const { container } = render(<SearchAutoComplete />);
  expect(container).toMatchSnapshot();
});

test("it should show the placeholder text", () => {
  const PLACE_HOLDER = "Hello Dog Placeholder";
  render(<SearchAutoComplete placeholder={PLACE_HOLDER} />);
  const placeholderElement = screen.getByPlaceholderText(PLACE_HOLDER);
  expect(placeholderElement).toBeInTheDocument();
});

test("it should be auto-focused on render", () => {
  render(<SearchAutoComplete autofocus={true} />);
  const inputElement = screen.getByRole("textbox");
  expect(inputElement).toHaveFocus();
});

test("it should not be auto-focused on render", () => {
  render(<SearchAutoComplete autofocus={false} />);
  const inputElement = screen.getByRole("textbox");
  expect(inputElement).not.toHaveFocus();
});

test("it should show 2 corgi items", async () => {
  render(<SearchAutoComplete />);
  const SEARCH = "corgi";
  const inputElement = screen.getByRole("textbox") as HTMLInputElement;
  expect(inputElement).toBeInTheDocument();

  fireEvent.change(inputElement, { target: { value: SEARCH } });
  expect(inputElement.value).toBe(SEARCH);

  //Search DOM for all elements with 'Welsh Corgi' (case insensitive)
  const items = await screen.findAllByText(/\bwelsh\s+corgi\b/i);
  expect(items).toHaveLength(2);
});

test("it should call useRouter when suggestion is clicked", async () => {
  const routerSpy = jest.spyOn(require("next/navigation"), "useRouter");
  render(<SearchAutoComplete />);

  const SEARCH = "corgi";
  const SUGGESTION = "Pembroke Welsh Corgi";
  const inputElement = screen.getByRole("textbox") as HTMLInputElement;
  expect(inputElement).toBeInTheDocument();

  fireEvent.change(inputElement, { target: { value: SEARCH } });
  expect(inputElement.value).toBe(SEARCH);

  const items = await screen.findByText(SUGGESTION);

  fireEvent.click(items);
  expect(inputElement.value).toBe(SUGGESTION);

  //Router is called once on render, another on suggestion click
  expect(routerSpy).toHaveBeenCalledTimes(2);
});

test("it should call onSubmit on userInput enter", async () => {
  render(<SearchAutoComplete />);

  const handleOnSubmitMock = jest.fn();
  const formElement = screen.getByRole("form");
  formElement.onsubmit = handleOnSubmitMock;

  const SEARCH = "corgi";
  const inputElement = screen.getByRole("textbox") as HTMLInputElement;
  expect(inputElement).toBeInTheDocument();
  fireEvent.change(inputElement, { target: { value: SEARCH } });
  expect(inputElement.value).toBe(SEARCH);

  expect(handleOnSubmitMock).not.toHaveBeenCalled();

  fireEvent.submit(formElement);
  expect(handleOnSubmitMock).toHaveBeenCalled();
});
