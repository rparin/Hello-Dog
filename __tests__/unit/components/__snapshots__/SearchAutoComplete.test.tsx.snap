// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders component unchanged 1`] = `
<div>
  <form
    aria-label="form"
  >
    <label
      aria-label="Dog Breed Search Bar"
    >
      <div
        class="sc-dcJsrY kdeOcX w-full"
      >
        <div
          class="wrapper"
        >
          <div
            class="sc-fqkvVR dYrfwz"
          >
            <input
              data-test="search-input"
              placeholder=""
              spellcheck="false"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </label>
  </form>
</div>
`;
